### PR TITLE
Rework ADC module to follow 'Extended API Guideline'

### PR DESCRIPTION
**Pull request by hs0225**
Tested on Artik10 and Stm32f4-discovery

IoT.js-DCO-1.0-Signed-off-by: Hosung Kim hs852.kim@samsung.com

### CODE REVIEW

File: src/js/adc.js [hunk 1]
```diff
@@ -17,42 +17,90 @@ var util = require('util');
...
-    if (util.isFunction(callback)) {
-      callback(err);
+function AdcPin(configuration, callback) {
+  var self = this;
+
+  if (util.isObject(configuration)) {
+    if (process.platform === 'linux') {
+      if (!util.isString(configuration.device)) {
+        throw new TypeError(
+          'Bad configuration - device is mandatory and String');
```
Comment (chokobole) [inline src/js/adc.js]: I think error message should be like `device is mandatory and should be String`. Current message right now a little bit confuses me that device is now type of String.

File: src/js/adc.js [hunk 2]
```diff
@@ -17,42 +17,90 @@ var util = require('util');
...
+      }
+    } else if (process.platform === 'nuttx') {
+      if (!util.isNumber(configuration.pin)) {
+        throw new TypeError('Bad configuration - pin is mandatory and Number');
```
Comment (chokobole) [inline src/js/adc.js]: ditto

File: src/module/iotjs_module_adc.c [hunk 3]
```diff
@@ -152,107 +150,137 @@ void iotjs_adc_after_work(uv_work_t* work_req, int status) {
...
-  const iotjs_jval_t* jcallback = JHANDLER_GET_ARG(1, function);
+  if (value < 0) {
+    req_data->result = false;
+    return;
+  }
 
-  iotjs_adc_reqwrap_t* req_wrap =
-      iotjs_adc_reqwrap_create(jcallback, kAdcOpExport);
+  req_data->value = value;
+  req_data->result = true;
```
Comment (chokobole) [inline src/module/iotjs_module_adc.c]: What do you think comparing this with below? IMHO, fast return is not proper to this case.
```c++
if (value > 0) {
  req_data->result = true;
  req_data->value = value;
} else {
  req_data->result = false;
}
```

File: src/module/iotjs_module_adc.c [hunk 4]
```diff
@@ -152,107 +150,137 @@ void iotjs_adc_after_work(uv_work_t* work_req, int status) {
...
+  ADC_WORKER_INIT;
 
-  iotjs_jhandler_return_null(jhandler);
+  // Release driver
+  if (!iotjs_adc_close(adc)) {
+    req_data->result = false;
+    return;
+  }
+
+  req_data->result = true;
```
Comment (chokobole) [inline src/module/iotjs_module_adc.c]: ditto
```c++
if (iotjs_adc_close(adc)) {
  req_data->result = true;
} else {
  req_data->result = false;
}
```

File: src/platform/iotjs_module_adc-linux-general.inl.h [hunk 5]
```diff
@@ -37,62 +35,37 @@
...
 
-void iotjs_adc_unexport_worker(uv_work_t* work_req) {
-  iotjs_adc_reqwrap_t* req_wrap = iotjs_adc_reqwrap_from_request(work_req);
-  iotjs_adc_reqdata_t* req_data = iotjs_adc_reqwrap_data(req_wrap);
-  DDDLOG("ADC %s()", __func__);
-  req_data->result = kAdcErrOk;
+  DDDLOG("%s()", __func__);
+  const char* device_path = iotjs_string_data(&_this->device);
+
+  // Check if ADC interface exits.
```
Comment (chokobole) [inline src/platform/iotjs_module_adc-linux-general.inl.h]: is this typo of `exists`?

File: src/platform/arm-nuttx/iotjs_module_adc-arm-nuttx.c [hunk 6]
```diff
@@ -79,73 +74,60 @@ static bool iotjs_adc_read_data(int32_t pin, struct adc_msg_s* msg) {
...
+    req_data->result = false;
     return;
   }
 
-  iotjs_gpio_unconfig_nuttx(pin);
+  DDLOG("%s - path: %s, number: %d, timer: %d", __func__, path, adc_number,
+        timer);
 
-  req_data->result = kAdcErrOk;
+  req_data->result = true;
```
Comment (chokobole) [inline src/platform/arm-nuttx/iotjs_module_adc-arm-nuttx.c]: ditto
```c++
if (adc_register(path, adc_dev) == 0) {
  req_data->result = true;
  DDLOG("%s - path: %s, number: %d, timer: %d", __func__, path, adc_number, timer);
} else {
  req_data->result = false;
}
```

File: src/platform/arm-nuttx/iotjs_module_adc-arm-nuttx.c [hunk 7]
```diff
@@ -79,73 +74,60 @@ static bool iotjs_adc_read_data(int32_t pin, struct adc_msg_s* msg) {
...
-    return -1;
+  // Release driver
+  if (unregister_driver(path) < 0) {
+    return false;
   }
 
-  return msg.am_data;
+  iotjs_gpio_unconfig_nuttx(pin);
+
+  return true;
```
Comment (chokobole) [inline src/platform/arm-nuttx/iotjs_module_adc-arm-nuttx.c]: ditto

File: src/module/iotjs_module_adc.h [hunk 8]
```diff
@@ -49,34 +53,30 @@ typedef struct {
...
 
 
-// This Adc class provides interfaces for ADC operation.
-typedef struct {
-  iotjs_jobjectwrap_t jobjectwrap;
-} IOTJS_VALIDATED_STRUCT(iotjs_adc_t);
+#define ADC_WORKER_INIT                                                     \
+  iotjs_adc_reqwrap_t* req_wrap = iotjs_adc_reqwrap_from_request(work_req); \
+  iotjs_adc_reqdata_t* req_data = iotjs_adc_reqwrap_data(req_wrap);         \
+  iotjs_adc_t* adc = iotjs_adc_instance_from_reqwrap(req_wrap);
```
Comment (chokobole) [inline src/module/iotjs_module_adc.h]: please wrap this with `do .. while(0)`

File: src/module/iotjs_module_adc.c (hunk 3 above)
Comment (hs0225) [inline src/module/iotjs_module_adc.c]: In my opinion, it is better to handle error first with an if statement.
In this code, if `value` is less than 0, it is exceptions.

File: src/module/iotjs_module_adc.c (hunk 4 above)
Comment (hs0225) [inline src/module/iotjs_module_adc.c]: ditto

File: src/platform/arm-nuttx/iotjs_module_adc-arm-nuttx.c (hunk 6 above)
Comment (hs0225) [inline src/platform/arm-nuttx/iotjs_module_adc-arm-nuttx.c]: ditto

File: src/module/iotjs_module_adc.h (hunk 8 above)
Comment (hs0225) [inline src/module/iotjs_module_adc.h]: The variables declared here are still used outside `define` directives. If these variables are declared in a loop, it is no longer available.

File: src/platform/arm-nuttx/iotjs_module_adc-arm-nuttx.c (hunk 7 above)
Comment (hs0225) [inline src/platform/arm-nuttx/iotjs_module_adc-arm-nuttx.c]: ditto

File: src/module/iotjs_module_adc.h (hunk 8 above)
Comment (chokobole) [inline src/module/iotjs_module_adc.h]: Ah okay!

File: src/module/iotjs_module_adc.c (hunk 3 above)
Comment (chokobole) [inline src/module/iotjs_module_adc.c]: With this philosophy which is `handle exception first`, then  I agreed

**Comment by chokobole** [comment]
LGTM

**Comment by yichoi** [comment]
LGTM

**Pull request merged**